### ISSUE TITLE
Shift towards an OAuth flow for API access

### ISSUE THREAD

**Issue by tienne-B** [TabbycatDebate/tabbycat#2350]
Having an OAuth flow for linking Tabbycat instances to add-ons would increase security and UX by not having what is essentially a password being re-used, and can have the apps request the specific permissions that they'd need. It would also be a smoother process for users who'd not need to find / copy their keys, rather being redirected to accept the request.

**Comment by tienne-B** [comment]
> allow Tabbycat to act as an OAuth provider

Correct! Another thing to point out is that we should be able to use the `Permission` enum in `users/permissions.py` as scopes, even if they can't be restricted by tournament.

**Comment by tienne-B** [comment]
Looks like a good library!

I'd recommend using the existing list of permissions as scopes. I see OAuth as primarily being for users to use external tools such as Tabtastic, where we'd want to tell them specifically what the integrations will do (and the end-user doesn't have to debug). When devs are working it through, we could add a list "required scopes" to the docs for each endpoint, or (inclusive) in 401 errors.

**Comment by teymour-aldridge** [comment]
Unfortunately I kind of got stuck here – I got a bit stuck working out how to get the subclassing to work. I will probably try to pick this up again in the future, but if anyone else is free in the meantime feel free to pick this up instead of me.
